src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/HierarchicalLayoutManager.java

```diff
@@ -23,1700 +23,723 @@
-        new BuildDatastructure().start();
-
-        if (!layoutSelfEdges) {
-            // Remove self-edges from the beginning.
-            removeSelfEdges(false);
-        }
-
-        // #############################################################
-        // STEP 2: Reverse edges, handle backedges
-        new ReverseEdges().start();
-
-        for (LayoutNode n : nodes) {
-            ArrayList<LayoutEdge> tmpArr = new ArrayList<>();
-            for (LayoutEdge e : n.succs) {
-                if (importantLinks.contains(e.link)) {
-                    tmpArr.add(e);
-                }
-            }
-
-            for (LayoutEdge e : tmpArr) {
-                e.from.succs.remove(e);
-                e.to.preds.remove(e);
-            }
-        }
-
-        // Hide self-edges from the layout algorithm and save them for later.
-        removeSelfEdges(true);
+    public void doLayout(LayoutGraph layoutGraph) {
+        layoutGraph.initializeLayout();
 
-        // #############################################################
-        // STEP 3: Assign layers
-        new AssignLayers().start();
+        // STEP 1: Remove self edges and reverse edges
+        ReverseEdges.apply(layoutGraph);
 
-        // #############################################################
-        // STEP 4: Create dummy nodes
-        new CreateDummyNodes().start();
+        // STEP 2: Assign layers and create dummy nodes
+        LayerManager.apply(layoutGraph, maxLayerLength);
 
-        // #############################################################
-        // STEP 5: Crossing Reduction
-        new CrossingReduction().start();
+        // STEP 3: Crossing Reduction
+        CrossingReduction.apply(layoutGraph);
 
-        // #############################################################
-        // STEP 7: Assign X coordinates
-        new AssignXCoordinates().start();
+        // STEP 4: Assign X coordinates
+        AssignXCoordinates.apply(layoutGraph);
 
-        // #############################################################
-        // STEP 6: Assign Y coordinates
-        new AssignYCoordinates().start();
-
-        // Put saved self-edges back so that they are assigned points.
-        for (LayoutEdge e : selfEdges) {
-            e.from.succs.add(e);
-            e.to.preds.add(e);
-        }
+        // STEP 5: Write back to interface
+        WriteResult.apply(layoutGraph);
 
-        // #############################################################
-        // STEP 8: Write back to interface
-        new WriteResult().start();
+        graph = layoutGraph;
     }
 
-    private class WriteResult extends AlgorithmPart {
+    public List<LayoutNode> getNodes() {
+        return graph.getAllNodes();
+    }
 
-        private int pointCount;
+    public static class ReverseEdges {
 
-        @Override
-        protected void run() {
+        static public void apply(LayoutGraph graph) {
+            removeSelfEdges(graph);
```

Comment:
The proposed code removes self-edges unconditionally. This is OK for the sea-of-nodes layout, but for the CFG layout we do need to draw self-edges (think about single basic block loops). Here is an (artificially edited) example of how the new algorithm misses drawing a self-edge for B7 (left is current IGV, right is IGV with your proposed changes):

![Screenshot from 2024-11-28 10-09-09](https://github.com/user-attachments/assets/e91b55af-0fbf-4f28-b7e6-558dfcada42a)

Here is the artificially edited graph file that illustrates the issue: [self-edges.zip](https://github.com/user-attachments/files/17945729/self-edges.zip)

Reply:
should be fixed now